### PR TITLE
Tests for WithdrawalQueue, share rate simulation fix

### PR DESCRIPTION
**Pull request by Jeday**
## Description
Update and expand existing tests to increase coverage and test edge cases.

## TODO:

- [x]  lido stub in existing test
- [x]  initialization (half of it already present)
- [x]  one request
- [x]  request amount boundaries
- [x]  request params checks
- [ ]  Multiple requests (variations?)
- [x]  finalization
- [x]  finalization params check
- [x]  multiple finalizations with same discount
- [ ]  multiple finalizations with different discounts
- [ ]  discount boundaries
- [ ]  100% discount !
- [x]  claim
- [x]  claim params check
- [x]  claim scenarios
    - [x]  direct order
    - [x]  reverse order
    - [x]  with holes
    - [x]  with different rates
    - [x]  fuzzing
- [x]  view methods
- [x]  find methods
    - [x]  direct
    - [x]  reverse
    - [x]  pseudorandom
- [x]  roles
- [x]  pause/resume
- [x]  nft
    - [x]  mint
    - [x]  burn

### CODE REVIEW

**Comment by folkyatina** [comment]
Guys, can you, please, rebase it to #648 if it's not so hard.

File: test/0.8.9/withdrawal-queue-deploy.test.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+      it('emits InitializedV1', async () => {
+        const { initTx } = await deployWithdrawalQueue({
+          stethOwner,
+          queueAdmin,
+          queuePauser,
+          queueResumer,
+          queueFinalizer,
+          queueBunkerReporter
+        })
+        await assert.emits(initTx, 'InitializedV1', {
```
Comment (folkyatina) [inline test/0.8.9/withdrawal-queue-deploy.test.js:78]: emits is not async, BTW

File: test/0.8.9/withdrawal-queue.test.js [hunk 2]
```diff
@@ -33,7 +33,7 @@ contract('WithdrawalQueue', ([owner, stranger, daoAgent, user]) => {
     await steth.approve(withdrawalQueue.address, StETH(300), { from: user })
 
     await impersonate(ethers.provider, steth.address)
```
Comment (skozin) [inline test/0.8.9/withdrawal-queue.test.js:35]: maybe let's just use a separate address for finalization as I've done in `withdrawal-queue-share-rate-changes.test.js`? it's anyway role-based so using stETH mock doesn't add any value here

File: test/0.8.9/withdrawal-queue.test.js [hunk 3]
```diff
@@ -1,30 +1,30 @@
...
+    })
 
-    await withdrawalQueue.initialize(daoAgent, daoAgent, daoAgent, steth.address, steth.address)
-    await withdrawalQueue.resume({ from: daoAgent })
+    steth = deployed.steth
+    wsteth = deployed.wsteth
+    withdrawalQueue = deployed.withdrawalQueue
 
     await steth.setTotalPooledEther(ETH(600))
     // we need 1 ETH additionally to pay gas on finalization because coverage ingnores gasPrice=0
```
Comment (skozin) [inline test/0.8.9/withdrawal-queue.test.js:30]: if using a separate finalizer, this extra ETH on the stETH address won't be needed

Comment (Jeday) [reply test/0.8.9/withdrawal-queue.test.js:30]: Yeah, that makes sense. We have talked about this but decided to keep things as is while we are working on additional tests so not to break existing ones.  
When finalizing this PR, I will see if we can sort it out.

Comment (folkyatina) [reply test/0.8.9/withdrawal-queue.test.js:30]: Yeah, good idea, basically.

**Comment by TheDZhon** [comment]
Merged the latest changes from `[fix/shapella-upgrade-fixes](https://github.com/lidofinance/lido-dao/tree/fix/shapella-upgrade-fixes) to simplify the diff.

**Review by TheDZhon — APPROVED** [review]
LGTM :+1:
